### DISCUSSION TITLE
Still failed to deploy milvus cluster with k8s offline

### DISCUSSION THREAD

**Discussion by AliscaCL** [milvus-io/milvus#16062] [Q&A and General discussion] [answered]
Hello, guys. Sorry to interrupt you again.

I have tried sevral time for deploy milvus cluster in offline k8s environment. 
Currently, Only minio pods has been successfully running. But others don't.

The result info for "kubectl get pods" is :
![微信图片_20220316092921](https://user-images.githubusercontent.com/96522752/158498789-51997109-99d3-409e-815a-8428107bc425.jpg)

The events logs for "kubectl get events" is :
![微信图片_20220316092925](https://user-images.githubusercontent.com/96522752/158498820-e8ccdfce-20eb-4a88-82b4-ff028f9ede03.jpg)

**Comment by LoveEachDay** [comment, marked as answer]
milvus cluster has three dependencies: etcd, minio and pulsar. 
1. etcd is responsible for service discovery and metadata storage. And etcd cluster must be ready before milvus service can normally startup.
2. minio is responsible for collection and index data storage. And minio cluster must be ready before milvus service can normally startup.
3. pulsar is responsible for request processing like create collection, insert data into collection. And pulsar cluster must be ready before milvus service can normally startup.

And pulsar cluster has 4 main components: zookeeper, bookkeeper, broker, and proxy. And the start order is zookeeper --> zookeeper metadata --> bookkeeper-->broker-->proxy. 
pulsar zookeeper is a statefulset, the start order is zookeeper-0--->zookeeper-1-->zookeeper-2.

From your screenshots above, we can see the pulsar cluster and etcd cluster is not ready. You should first check the log of zookeeper-0 to find why the pod is not ready.

**Reply by AliscaCL** [reply]
Yes, I guess that. 
Currently the event message for zookeeper-0 is shown as follows:

Successfully assigned opinion/opinion-release-pulsar-zookeeper-0 to tdh03
Container image "apachepulsar/pulsar:2.8.2" already present on machine
Container opinion-release-pulsar-zookeeper failed liveness probe, will be restared
Started container opinion-release-pulsar-zookeeper
Liveness probe failed: ZK server is not ok
Readiness probe failed: ZK server is not ok
Back-off restarting failed container.

What is the ZK server? What should I do for checking ZK server?

**Reply by AliscaCL** [reply]
T-T I think the result was found by myself....
It is a really silly issue. The k8s too slow or the readiness time is too short. Just raise the time limit for readiness time will be work.

**Comment by LoveEachDay** [comment]
ZK server is short for `zookeeper server`. You need to check the zookeeper pod log not just the events.

**Reply by AliscaCL** [reply]
Hello~
The logs show that
secureClientPort is not set
observerMasterPort is not set
![微信图片_20220316134946 (2)](https://user-images.githubusercontent.com/96522752/158527396-655e58ae-3175-471f-ae70-81a1dbc5eb9a.png)

Where should I set these port?

**Reply by AliscaCL** [reply]
T-T Do you have any idea?

**Reply by AliscaCL** [reply]
Here is the actually logs for zookeeper error.
Nothing report....

**Comment by LoveEachDay** [comment]
Would you provide your installation manifests?

**Reply by AliscaCL** [reply]
Thank you so much.
The manifest is already in a offline-env, it is hard to extract it from that environment. 
But thank you. Base on your suggestion, I have check what will happened on mannual start up zookeeper server. I found the logs show as your next response. Then I realized that, it might caused by the health-machinesim for k8s. It will check it every 10 sec. But the server haven't start up in my cluster. That is why it will restart hundreds of times.

**Comment by LoveEachDay** [comment]
Here is a normal zookeeper startup log:
```
[conf/zookeeper.conf] Adding config quorumListenOnAllIPs = true
[conf/pulsar_env.sh] Applying config JAVA_HOME = /usr/lib/jvm/java-11-openjdk-amd64
[conf/pulsar_env.sh] Applying config PULSAR_EXTRA_OPTS = -Dpulsar.log.root.level=info
[conf/pulsar_env.sh] Applying config PULSAR_GC = "-XX:+UseG1GC"
[conf/pulsar_env.sh] Applying config PULSAR_MEM = "-Xms1024m -Xmx1024m -Dcom.sun.management.jmxremote -Djute.maxbuffer=10485760 -XX:+DoEscapeAnalysis -XX:+DisableExplicitGC -XX:+PerfDisableSharedMem -Dzookeeper.forceSync=no"
[conf/pulsar_env.sh] Adding config quorumListenOnAllIPs = true
Current server id 1
Creating data/zookeeper/myid with id = 1
11:14:15.510 [main] INFO  org.apache.zookeeper.server.quorum.QuorumPeerConfig - Reading configuration from: /pulsar/conf/zookeeper.conf
11:14:15.514 [main] WARN  org.apache.zookeeper.server.quorum.QuorumPeerConfig - data/zookeeper is relative. Prepend ./ to indicate that you're sure!
11:14:15.515 [main] INFO  org.apache.zookeeper.server.quorum.QuorumPeerConfig - clientPortAddress is 0.0.0.0:2181
11:14:15.515 [main] INFO  org.apache.zookeeper.server.quorum.QuorumPeerConfig - secureClientPort is not set
11:14:15.515 [main] INFO  org.apache.zookeeper.server.quorum.QuorumPeerConfig - observerMasterPort is not set
11:14:15.515 [main] INFO  org.apache.zookeeper.server.quorum.QuorumPeerConfig - metricsProvider.className is org.apache.zookeeper.metrics.prometheus.PrometheusMetricsProvider
11:14:15.531 [main] INFO  org.apache.zookeeper.server.DatadirCleanupManager - autopurge.snapRetainCount set to 3
11:14:15.531 [main] INFO  org.apache.zookeeper.server.DatadirCleanupManager - autopurge.purgeInterval set to 1
11:14:15.532 [PurgeTask] INFO  org.apache.zookeeper.server.DatadirCleanupManager - Purge task started.
11:14:15.533 [main] INFO  org.apache.zookeeper.jmx.ManagedUtil - Log4j 1.2 jmx support not found; jmx disabled.
11:14:15.533 [main] INFO  org.apache.zookeeper.server.quorum.QuorumPeerMain - Starting quorum peer, myid=1
11:14:15.537 [PurgeTask] INFO  org.apache.zookeeper.server.persistence.FileTxnSnapLog - zookeeper.snapshot.trust.empty : true
11:14:15.543 [PurgeTask] INFO  org.apache.zookeeper.server.DatadirCleanupManager - Purge task completed.
11:14:15.546 [main] INFO  org.apache.zookeeper.metrics.prometheus.PrometheusMetricsProvider - Initializing metrics, configuration: {exportJvmInfo=true, httpPort=8000}
11:14:15.547 [main] INFO  org.apache.zookeeper.metrics.prometheus.PrometheusMetricsProvider - Starting /metrics HTTP endpoint at port 8000 exportJvmInfo: true
11:14:15.571 [main] INFO  org.eclipse.jetty.util.log - Logging initialized @945ms to org.eclipse.jetty.util.log.Slf4jLog
11:14:15.640 [main] INFO  org.eclipse.jetty.server.Server - jetty-9.4.43.v20210629; built: 2021-06-30T11:07:22.254Z; git: 526006ecfa3af7f1a27ef3a288e2bef7ea9dd7e8; jvm 11.0.13+8-Ubuntu-0ubuntu1.20.04
11:14:15.672 [main] INFO  org.eclipse.jetty.server.handler.ContextHandler - Started o.e.j.s.ServletContextHandler@28a0fd6c{/,null,AVAILABLE}
11:14:15.687 [main] INFO  org.eclipse.jetty.server.AbstractConnector - Started ServerConnector@2f9a01c1{HTTP/1.1, (http/1.1)}{0.0.0.0:8000}
11:14:15.687 [main] INFO  org.eclipse.jetty.server.Server - Started @1064ms
```

**Reply by AliscaCL** [reply]
Thank you ~

**Reply by AliscaCL** [reply]
The zookeeper seems ready. 
![zookeeper_ready](https://user-images.githubusercontent.com/96522752/158734568-efeee384-5661-4935-961b-48af2c8d0687.jpg)

But zookeeper-meta-data and other pods still tells me that they are waiting for zookeeper.
![still_wait](https://user-images.githubusercontent.com/96522752/158734632-11d8cb34-a0fd-4d20-a926-16da1fc3ce24.jpg)

Is that correct? The only thing I should do is waiting?

**Comment by LoveEachDay** [comment]
wait-zookeeper-ready just check for `zookeeper-2` has a dns lookup. Can you check whether the coredns service in `kube-system` namespace behaves correctly.

**Reply by AliscaCL** [reply]
How to do it?

**Comment by LoveEachDay** [comment]
1. check the dns service has any endpoints:
`kubectl describe svc kube-dns -n kube-system`
2. if step 1 says the `kube-dns` has endpoints, check the endpoint pods:
`kubectl get pods -n kube-system | grep coredns`
3. check the log of the pod

**Reply by AliscaCL** [reply]
For cmd: kubectl describe svc kube-dns -n kube-system
The endpoint shows:
170.0.51.145:53, 170.0.51.146:53, 170.0.51.147:53

But for cmd kubectl get pods -n kube-system | grep coredns 
the result is missing.

The coorelated dns pods, names:
kube-dns-flannel

**Reply by AliscaCL** [reply]
But I can try it in another cluster.

**Comment by LoveEachDay** [comment]
Maybe you can try `kubectl get pods -n kube-system | grep dns` instead.

**Reply by AliscaCL** [reply]
The result shows:
kube-dns-flannel-j42zp   1/1 Running 0 54d
kube-dns-flannel-m6x72 1/1 Running 0 54d
kube-dns-flannel-zpnnx  1/1 Running 0 54d

**Comment by LoveEachDay** [comment]
Check the pod log, see any request of `opinion-release-pulsar-zookeeper-2`

**Reply by AliscaCL** [reply]
![Uploading zookeeper-log.jpg…]()
Hello, Here is the result.

**Reply by AliscaCL** [reply]
Here is the result.
![zookeeper-log](https://user-images.githubusercontent.com/96522752/158785081-c3e09cb9-db0d-4039-b0d3-1758c1263bc1.jpg)

**Reply by LoveEachDay** [reply]
![image](https://user-images.githubusercontent.com/1573213/158788364-74e72cf5-76fc-47a7-b814-b2666b24ef3b.png)
Seems your dns lookup doesn't behave normally.

**Reply by LoveEachDay** [reply]
Contact your kubernetes cluster administrator to fix.

**Reply by AliscaCL** [reply]
A little bad news is, the administrator is me.... 😢

**Comment by AliscaCL** [comment]
I have check the network configuration.
The kube-dns address is :
![dns-ip](https://user-images.githubusercontent.com/96522752/158925263-ff0f6637-ecd2-490c-99cd-5c410e4d3a5d.jpg)

and the dns config in zookeeper pods is:
![dns-service](https://user-images.githubusercontent.com/96522752/158925319-54ca86d1-d48d-45ef-a578-1145d69c39e7.jpg)

The pod ip address is:
![pod-ip](https://user-images.githubusercontent.com/96522752/158925365-bd975087-8df4-49fa-a979-cf8b43d2a5f5.jpg)

Do you have any idea?
